rasa/core/migrate.py

```diff
@@ -238,59 +262,114 @@ def _migrate_domain_files(
+    domain_path = Path(domain_path)
+    out_path = Path(out_path) if out_path else None
+
+    domain_parent_dir = domain_path.parent
+    migrate_file_only = domain_path.is_file()
+
+    # Ensure the backup location does not exist yet
+    # Note: We demand that file as well as folder with this name gets deleted before
+    # the command is run to avoid confusion afterwards.
+    suffix = "original_domain"
+    suffix = f"{suffix}.yml" if migrate_file_only else suffix
+    backup_location = domain_parent_dir / suffix
+    if backup_location.exists():
+        backup_location_str = "directory" if backup_location.isdir() else "file"
+        raise RasaException(
+            f"The domain from '{domain_path}' could not be migrated since the "
+            f"a {backup_location_str} {backup_location} already exists."
+            f"Please remove that there is no file or folder at {backup_location}."
+        )
 
-    if domain_file.is_dir():
-        backup_location = current_dir / "original_domain"
-        backup_location.mkdir()
-
-        if out_file.is_file() or not out_file.exists():
-            out_file = current_dir / "new_domain"
-            out_file.mkdir()
-            created_out_dir = True
-            rasa.shared.utils.io.raise_warning(
-                f"The out path provided is not a directory, "
-                f"creating a new directory '{str(out_file)}' "
-                f"for migrated domain files."
-            )
+    # Choose a default output location if nothing was specified
+    if out_path is None:
+        suffix = DEFAULT_DOMAIN_PATH if migrate_file_only else "new_domain"
+        out_path = domain_parent_dir / suffix
 
-        try:
-            original_domain = _migrate_domain_files(
-                domain_file, backup_location, out_file
+    # Ensure the output location is not already in-use
+    if not migrate_file_only:
+        if out_path.is_dir() and any(out_path.iterdir()):
+            raise RasaException(
+                f"The domain from '{domain_path}' could not be migrated to "
+                f"{out_path} because that folder is not empty."
+                "Please remove the folder and try again."
             )
-        except Exception as e:
-            shutil.rmtree(backup_location)
-            if out_file != domain_file:
-                shutil.rmtree(out_file)
-                if not created_out_dir:
-                    # we recreate the deleted directory
-                    # because it existed before
-                    out_file.mkdir()
-            raise e
     else:
-        if not Domain.is_domain_file(domain_file):
+        if out_path.is_file():
             raise RasaException(
-                f"The file '{domain_file}' could not be validated as a "
-                f"domain file. Only domain yaml files can be migrated. "
+                f"The domain from '{domain_path}' could not be migrated to "
+                f"{out_path} because a file already exists."
```

Comment:
```suggestion
                f"'{out_path}' because a file already exists."
```
Plus I think this might need a unit test too?

Reply:
done